foundationdb/src/future.rs

```diff
@@ -0,0 +1,117 @@
+use std;
+
+use foundationdb_sys as fdb;
+use futures;
+use futures::Async;
+
+use error::{FdbError, Result};
+
+pub struct FdbFuture {
+    //
+    f: *mut fdb::FDBFuture,
+    task: Option<Box<futures::task::Task>>,
+}
+
+impl FdbFuture {
+    pub(crate) fn new(f: *mut fdb::FDBFuture) -> Self {
+        Self { f, task: None }
+    }
+}
+
+impl futures::Future for FdbFuture {
+    type Item = FdbFutureResult;
+    type Error = FdbError;
+
+    fn poll(&mut self) -> std::result::Result<Async<Self::Item>, Self::Error> {
+        if self.task.is_none() {
+            let task = futures::task::current();
+            let task = Box::new(task);
+            let task_ptr = task.as_ref() as *const _;
+            unsafe {
+                fdb::fdb_future_set_callback(self.f, Some(fdb_future_callback), task_ptr as *mut _);
+            }
+            self.task = Some(task);
+
+            return Ok(Async::NotReady);
+        }
+
+        let ready = unsafe { fdb::fdb_future_is_ready(self.f) };
+        if ready == 0 {
+            return Ok(Async::NotReady);
+        }
+
+        let err = unsafe { fdb::fdb_future_get_error(self.f) };
+        if err != 0 {
+            return Err(FdbError::from(err));
+        }
+
+        let g = FdbFutureResult::new(self.f);
```

Comment:
let's add a comment here that FdbFutureResult is taking ownership of the FDBFuture

Reply:
I added a comment.